### PR TITLE
Add array_frequency Presto function

### LINKED ISSUE #3752: Add array_frequency Presto function
### Description

Implement a [scalar function](https://facebookincubator.github.io/velox/develop/scalar-functions.html) of presto's equivalent [array_frequency](https://prestodb.io/docs/current/functions/array.html#array_frequency)
Presto implementation at [ArraySqlFunctions](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/ArraySqlFunctions.java#L58,81)

### PR DESCRIPTION
**Pull request by vermapratyush**
Summary:
### Test Plan

```
$ velox_expression_fuzzer_test --  --seed 100 --lazy_vector_generation_ratio 0.2 --duration_sec 600 --enable_variadic_signatures --velox_fuzzer_enable_complex_types --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --retry_with_try --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/fuzzer_repro --only array_frequency

I0126 12:37:10.402107 2149293 ExpressionFuzzer.cpp:949] ==============================> Done with iteration 22412
I0126 12:37:10.402221 2149293 ExpressionFuzzer.cpp:829] ==============================> Top 1 by number of rows processed
I0126 12:37:10.402263 2149293 ExpressionFuzzer.cpp:831] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0126 12:37:10.402277 2149293 ExpressionFuzzer.cpp:835] array_frequency 22413 100.00% 340170
I0126 12:37:10.402318 2149293 ExpressionFuzzer.cpp:841] ==============================> Bottom 1 by number of rows processed
I0126 12:37:10.402338 2149293 ExpressionFuzzer.cpp:843] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0126 12:37:10.402352 2149293 ExpressionFuzzer.cpp:848] array_frequency 22413 100.00% 340170
I0126 12:37:10.402372 2149293 ExpressionFuzzer.cpp:861] ==============================> All stats sorted by number of times the function was chosen
I0126 12:37:10.402385 2149293 ExpressionFuzzer.cpp:863] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0126 12:37:10.402405 2149293 ExpressionFuzzer.cpp:867] array_frequency 22413 100.00% 340170

```

Fuzzer on all functions
```
$ velox_expression_fuzzer_test --  --seed 100 --lazy_vector_generation_ratio 0.2 --duration_sec 600 --enable_variadic_signatures --velox_fuzzer_enable_complex_types --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --retry_with_try --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/fuzzer_repro

https://pastebin.com/1k4cGMhj
```

Fixes #3752

Differential Revision: D42712223

### CODE REVIEW

**Comment by vermapratyush** [comment]
Failures unrelated to the diff.

File: velox/functions/prestosql/tests/ArrayFrequencyTest.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+       1,
+       1,
+       // empty value
+       5,
+       1,
+       3,
+       2,
+       2,
+       2});
+  auto expected = makeMapVector({0, 3, 3, 5, 6}, expectedKeys, expectedValues);
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayFrequencyTest.cpp:73]: Any particular reason not to use

```
  // Create map vector from nested std::vector representation.
  template <typename TKey, typename TValue>
  MapVectorPtr makeMapVector(
      const std::vector<std::vector<std::pair<TKey, std::optional<TValue>>>>&
          maps,
      const TypePtr& type =
          MAP(CppToType<TKey>::create(), CppToType<TValue>::create())) {
```

**Review by mbasmanova — COMMENTED** [review]
@vermapratyush Thank you for adding one more Presto function. Would you update PR description with results of Fuzzer testing?

> Implement array_frequency in velox with same signature [as presto](https://prestodb.io/docs/current/functions/array.html#array_frequency).

This statement is redundant. Consider removing.

File: velox/functions/prestosql/ArrayFunctions.h [hunk 2]
```diff
@@ -399,4 +399,30 @@ struct ArrayHasDuplicatesFunction {
...
+  VELOX_DEFINE_FUNCTION_TYPES(TExecParams);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<velox::Map<T, int>>& out,
+      arg_type<velox::Array<T>> inputArray) {
+    folly::F14FastMap<arg_type<T>, int> frequencyCount;
+
+    // optimization to skip counting frequency of null values
+    if (inputArray.mayHaveNulls()) {
+      for (const auto& item : inputArray.skipNulls()) {
```
[the remark this answers is not in the document]
Comment (vermapratyush) [reply velox/functions/prestosql/ArrayFunctions.h:415]: @mbasmanova the API is different as `.skipNulls` also unrolls the value. Hence in the subsequent branch is 
```
frequencyCount[item.value()]++;
vs
frequencyCount[item]++;
```

File: velox/functions/prestosql/tests/ArrayFrequencyTest.cpp (hunk 1 above)
Comment (vermapratyush) [reply velox/functions/prestosql/tests/ArrayFrequencyTest.cpp:73]: somehow missed that api, definitely much cleaner

**Comment by vermapratyush** [comment]
@mbasmanova Running fuzzer without `--only` params failed with the following error.
Is this a known issue?
```
$ buck run //velox/expression/tests:velox_expression_fuzzer_test --  --seed 100 --lazy_vector_generation_ratio 0.2 --duration_sec 600 --enable_variadic_signatures --velox_fuzzer_enable_complex_types --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --retry_with_try --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/fuzzer_repr

terminate called after throwing an instance of 'facebook::velox::VeloxUserError'
  what():  Exception: VeloxUserError
Error Source: USER
Error Code: UNSUPPORTED
Reason: Unsupported map key type for element_at: INTERVAL DAY TO SECOND
Retriable: False
Function: applyMap
File: buck-out/dev/gen/aab7ed39/velox/functions/lib/velox_functions_lib#header-mode-symlink-tree-with-header-map,headers/velox/functions/lib/SubscriptUtil.h
Line: 144
Stack trace:
Stack trace has been disabled.Use --velox_exception_user_stacktrace=true to enable it.

*** Aborted at 1674764746 (Unix time, try 'date -d @1674764746') ***
*** Signal 6 (SIGABRT) (0x3227d0022efdd) received by PID 2289629 (pthread TID 0x7fc9898acf40) (linux TID 2289629) (maybe from PID 2289629, UID 205437) (code: -6), stack trace: ***
    @ 000000000001100e folly::symbolizer::(anonymous namespace)::innerSignalHandler(int, siginfo_t*, void*)
                       ./folly/experimental/symbolizer/SignalHandler.cpp:449
    @ 000000000000f731 folly::symbolizer::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)
                       ./folly/experimental/symbolizer/SignalHandler.cpp:470
    @ 000000000004459f (unknown)
    @ 000000000009c9d3 __GI___pthread_kill
    @ 00000000000444ec __GI_raise
    @ 000000000002c432 __GI_abort
    @ 00000000000a3fd4 __gnu_cxx::__verbose_terminate_handler()
    @ 00000000000a1b39 __cxxabiv1::__terminate(void (*)())
    @ 00000000000a1ba4 std::terminate()
    @ 00000000000a1ec1 __cxa_rethrow
    @ 0000000000022874 facebook::velox::test::ExpressionVerifier::verify(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, std::shared_ptr<facebook::velox::RowVector> const&, std::shared_ptr<facebook::velox::BaseVector>&&, bool, std::vector<unsigned int, std::allocator<unsigned int> >)
                       ./velox/expression/tests/ExpressionVerifier.cpp:190
    @ 000000000009a0f1 facebook::velox::test::ExpressionFuzzer::go()
                       ./velox/expression/tests/ExpressionFuzzer.cpp:941
    @ 00000000000c8ce6 facebook::velox::test::expressionFuzzer(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<facebook::velox::exec::FunctionSignature const*, std::allocator<facebook::velox::exec::FunctionSignature const*> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<facebook::velox::exec::FunctionSignature const*, std::allocator<facebook::velox::exec::FunctionSignature const*> > > > >, unsigned long)
                       ./velox/expression/tests/ExpressionFuzzer.cpp:957
    @ 00000000002e45cf FuzzerRunner::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
                       ./velox/expression/tests/FuzzerRunner.h:140
                       -> ./velox/expression/tests/ExpressionFuzzerTest.cpp
    @ 00000000002e3da7 main
                       ./velox/expression/tests/ExpressionFuzzerTest.cpp:67
    @ 000000000002c656 __libc_start_call_main
    @ 000000000002c717 __libc_start_main_alias_2
    @ 00000000002c3580 _start
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/csu/../sysdeps/x86_64/start.S:116
```

**Comment by mbasmanova** [comment]
@vermapratyush This is a known issue that was fixed in https://github.com/facebookincubator/velox/pull/3801

Would you rebase and retry?

**Review by mbasmanova — APPROVED** [review]
@vermapratyush Thanks.

**Pull request closed without merging**